sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,260 @@
+by any `barrier` with the specified `Scope` and `CompletionFunction`.
+
+[source,c++]
+----
+constexpr explicit barrier(ptrdiff_t expected, CompletionFunction f = CompletionFunction());
+----
+_Effects_: Initializes the barrier with the given expected count and completion
+function, and begins the first phase of the barrier.
+
+[source,c++]
+----
+[[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+----
+_Effects_: The calling thread of execution arrives at the barrier and decreases
+the expected count by `update`.
+
+_Returns_: An `arrival_token` that can be passed to `wait`.
+
+[source,c++]
+----
+void wait(arrival_token&& arrival) const;
+----
+_Effects_: Blocks the calling thread of execution until the end of the barrier
+phase associated with `arrival`.
+
+[NOTE]
+====
+Because calling `wait` blocks the calling thread of execution, it may prevent
+forward progress. Unlike `group_barrier`, the member functions of `barrier` do
+not provide additional scheduling guarantees; it is a user's responsibility to
+ensure that calls to `wait` are compatible with the forward progress guarantees
+provided by an implementation.
+====
+
+[source,c++]
+----
+void arrive_and_wait();
+----
+_Effects_: Equivalent to `wait(arrive())`.
+
+[source,c++]
+----
+void arrive_and_drop();
+----
+_Effects_: The calling thread of execution arrives at the barrier, decreases
+the number of threads of execution expected in the next phase, and decreases
+the expected count of the current phase by 1.
+
+
+=== `group_arrive` and `group_wait`
+
+This extension provides two convenience functions for `arrive` and `wait` with
+additional convergence requirements, to simplify reasoning about forward
+progress guarantees in common situations. Both of these functions act as
+synchronization points for all work-items in a group.
+
+[source,c++]
+----
+template <typename Group, typename Barrier>
+[[nodiscard]] group_arrival_token group_arrive(Group g, Barrier b);
+----
+_Effects_: Waits for all work-items in group `g` to reach this point of
+execution, then signals that all work-items have arrived at barrier `b` and
+decreases the expected count by `g.get_group_linear_range()`.
+
+_Returns_: A `group_arrival_token` that can be passed to `group_wait`.
+
+[NOTE]
+====
+Implementations may decrease the expected count via a call to `arrive(1)` from
+each work-item in the group, or via a single call to
+`arrive(g.get_group_linear_range())` from the elected leader of the group.
+====
+
+[source,c++]
+----
+template <typename Group>
+void group_wait(Group g, group_arrival_token&& arrival);
+----
+_Effects_: Waits for all work-items in group `g` to reach this point of
+execution, then blocks all work-items in group `g` until the end of the barrier
+phase associated with `arrival`.
+
+[NOTE]
+====
+Implementations may block the work-items in group `g` via a call to `wait` from
+each work-item in the group, or via a single call to `wait` from the elected
+leader of the group.
```

Comment:
The second option would only be valid if the sycl implementation this is running on executes workitems in the group `g` in lockstep.

Reply:
I don't think this is true.  My intent here was to say that because `group_wait` is a synchronization point for the group, an implementation like the one in the sketch below would be legal:

```c++
template <typename Group, typename Barrier>
void group_wait(Group g, Barrier b, group_arrival_token&& group_token) {
  sycl::group_barrier(g);
  if (g.leader()) {
    b.wait(detail::get_token_for_regular_wait(group_token));
  }
  sycl::group_barrier(g);
}
```

This wouldn't require the implementation to be running `g` in lockstep, but would require the implementation to introduce some synchronization.  Would adding a sketch like this to the implementation notes be helpful?

Note that I added the `Barrier` template argument here; while sketching this out, I realized we'd need that in order to perform the transformation I had envisioned.  I've added that in 3334cf37756fd27cb77c7551613d4e1497ce78a0.

Reply:
> My intent here was to say that because group_wait is a synchronization point for the group, an implementation like the one in the sketch below would be legal

In that case you should explicitly state that the group needs to be synchronized before and after the leader calls `wait` for this option to be a valid implementation. After you add that statement I do not think the sketched implementation is needed, but I am not against adding it either.

Reply:
I've tried to do this in 3aa93e9ac08d, by clarifying that `group_arrive` and `group_wait` are both group functions (as defined by the SYCL specification).  The part of the specification I'm referring to is copied below for convenience:

> These group functions act as synchronization points and must be encountered in converged [control flow](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#control-flow) by all work-items in the group. If one work-item in a group calls a group function, then all work-items in that group must call exactly the same function under the same set of conditions --- calling the same function under different conditions (e.g. in different iterations of a loop, or different branches of a conditional statement) results in undefined behavior. Additionally, restrictions may be placed on the arguments passed to each function in order to ensure that all work-items in the group agree on the operation that is being performed. Any such restrictions on the arguments passed to a function are defined within the descriptions of those functions. Violating these restrictions results in undefined behavior.

Do you agree that this covers it? The functions being synchronization points is defined elsewhere to mean that each work-item waits for all other work-items in the group to arrive before executing the function, and no work-item can move past the function until all other work-items have executed it.  I think that's equivalent to what you're suggesting, but without explicitly stating that any call to `wait` must have barriers on either side.

Reply:
I think it is going in the right direction, but I do not think that covers it. This specifies how a user should call the function, not how it should be implemented. So, this does not prescribe that there should be synchronization before and after wait if it is implemented by a single work item calling `wait`. So this note about potential implementation still needs to mention synchronization if a single work item calls `wait`.

Reply:
Ah, sorry, I didn't realize you were talking about the note(s).  I've updated them in 50d90faaff57.